### PR TITLE
Refactor item class [CR]

### PR DESCRIPTION
**Pull request by mugling**
This PR starts work on refactoring the `item` class
- Converts `item::make_corpse` into static member function
- Drop `item::init` in favor of C++11 member initialization
- Drop the union (`poision`, `bigness`, `frequency`, `note`, `irridation`)

### CODE REVIEW

File: src/item.cpp [hunk 1]
```diff
@@ -172,31 +172,25 @@ item::item(const std::string new_type, int turn, bool rand)
...
 }
 
-void item::make_corpse( const mtype_id& mt, unsigned int turn )
+item item::make_corpse( const mtype_id& mt, int turn, const std::string &name )
 {
     if( !mt.is_valid() ) {
         debugmsg( "tried to make a corpse with an invalid mtype id" );
     }
-    const bool isReviveSpecial = one_in( 20 );
-    init();
```
Comment (mugling) [inline src/item.cpp]: The goal is to remove this call. I'm uncertain if the one remaining use in `savegame_json.cpp` is necessary?

**Comment by kevingranade** [comment]
I'm ending up with "zombie corpse of human corpse" when I kill a zombie.
Trying to figure out why that's happening.

**Comment by kevingranade** [comment]
Zombie resurrection seems to work, it's just the corpse name that's weird.

**Comment by kevingranade** [comment]
The zombie corpse items are having their name field set to "human corpse".

File: src/item.cpp [hunk 2]
```diff
@@ -172,31 +171,25 @@ item::item(const std::string new_type, int turn, bool rand)
...
-    bday = turn;
-}
 
-void item::make_corpse( const mtype_id& mt, unsigned int turn, const std::string &name )
-{
-    make_corpse( mt, turn );
-    this->name = name;
-}
+    if( !name.empty() ) {
+        result.name = name;
```
Comment (mugling) [inline src/item.cpp]: We should provide an `else` statement setting `result.name = std::string()`

**Comment by mugling** [comment]
I'm going to compile that branch now and try that change

File: src/item.cpp [hunk 3]
```diff
@@ -99,12 +99,11 @@ static const itype *nullitem()
...
-    init();
+    type = nullitem();
 }
 
 item::item(const std::string new_type, int turn, bool rand)
 {
-    init();
     type = find_type( new_type );
     bday = turn;
     corpse = type->id == "corpse" ? &mtype_id::NULL_ID.obj() : nullptr;
```
Comment (kevingranade) [inline src/item.cpp]: corpse is being set to mon_null here, which has a name of "human corpse".

**Comment by mugling** [comment]
6019f00 seems to resolve it

File: src/item.cpp (hunk 2 above)
Comment (kevingranade) [inline src/item.cpp]: If name.empty() is true, then name is a blank string, so just use that?
That's the version I applied, and it'll take me longer to roll back my changes, add yours, and re-compile, so I'm going to go with my version :)
Feel free to change it later if you want.

File: src/item.cpp [hunk 4]
```diff
@@ -172,31 +171,25 @@ item::item(const std::string new_type, int turn, bool rand)
...
+    if( result.active && one_in( 20 ) ) {
+        result.item_tags.insert( "REVIVE_SPECIAL" );
     }
-    bday = turn;
-}
 
-void item::make_corpse( const mtype_id& mt, unsigned int turn, const std::string &name )
-{
-    make_corpse( mt, turn );
-    this->name = name;
```
Comment (kevingranade) [inline src/item.cpp]: BTW, this is where it was being reset before.

**Comment by kevingranade** [comment]
Pushed with an alternate final commit for convenience.

**Pull request closed without merging**